platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/EnhancedKeyStoreLoader.java

```diff
@@ -1165,4 +1171,326 @@ private interface AddressBookCallback {
+            logger.error(
+                    ERROR.getMarker(),
+                    "Failed to rollback {} pem files created. Manual cleanup required.",
+                    cleanupErrorCount.get());
+            throw new IllegalStateException("Cryptography Migration failed to generate or validate PEM files.");
+        }
+    }
+
+    /**
+     * Move the PFX files to the OLD_PFX_KEYS sub-directory.
+     *
+     * @throws KeyStoreException   if the underlying method calls throw this exception.
+     * @throws KeyLoadingException if the underlying method calls throw this exception.
+     */
+    private void cleanupMovePfxFilesToSubDirectory() throws KeyStoreException, KeyLoadingException {
+        AtomicLong cleanupErrorCount = new AtomicLong(0);
+        Path oldPfxKeys = keyStoreDirectory.resolve("OLD_PFX_KEYS");
+        if (!Files.exists(oldPfxKeys)) {
+            try {
+                Files.createDirectory(oldPfxKeys);
+            } catch (IOException e) {
+                logger.error(
+                        ERROR.getMarker(), "Failed to create OLD_PFX_KEYS sub-directory. Manual cleanup required.");
+                return;
+            }
+        }
+        iterateAddressBook(addressBook, (i, nodeId, address, nodeAlias) -> {
+            if (isLocal(address)) {
+                // move private key PFX files per local node
+                File sPrivatePfx = legacyPrivateKeyStore(nodeAlias).toFile();
+                if (sPrivatePfx.exists()
+                        && sPrivatePfx.isFile()
+                        && !sPrivatePfx.renameTo(
+                                oldPfxKeys.resolve(sPrivatePfx.getName()).toFile())) {
+                    cleanupErrorCount.incrementAndGet();
+                }
+            }
+        });
+        File sPublicPfx = legacyCertificateStore().toFile();
+        if (sPublicPfx.exists()
+                && sPublicPfx.isFile()
+                && !sPublicPfx.renameTo(oldPfxKeys.resolve(sPublicPfx.getName()).toFile())) {
+            cleanupErrorCount.incrementAndGet();
+        }
+        if (cleanupErrorCount.get() > 0) {
+            logger.error(
+                    ERROR.getMarker(),
+                    "Failed to move {} PFX files to OLD_PFX_KEYS. Manual cleanup required.",
+                    cleanupErrorCount.get());
+            throw new IllegalStateException("Cryptography Migration failed to move PFX files to OLD_PFX_KEYS sub-directory.");
+        }
+    }
+
+    /**
+     * Write the provided encoded key or certificate as a base64 DER encoded PEM file to the provided location.
+     *
+     * @param isPrivateKey true if the encoded data is a private key; false if it is a certificate.
+     * @param location     the location to write the PEM file.
+     * @param encoded      the byte encoded data to write to the PEM file.
+     * @throws IOException if an error occurred while writing the PEM file.
+     */
+    private static void writePemFile(
```

Comment:
Minor: just wondering (and I don't know the answer), doesn't JDK provide any utility methods for formatting PEM files? The logic is trivial, sure, but if there's a built-in utility, I'd rather delegate to it.

Reply:
I have [this](https://github.com/hashgraph/hedera-services/blob/7e2f3ca226f2e0878ac6645970a98dfbfe5db2f2/hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookHelper.java#L77) for writing certificate pem file if you want to do something similar.

Reply:
I am switching to the bouncy castle implementation.